### PR TITLE
[IMP] website_sale: automatic invoice for fully discounted carts

### PR DESCRIPTION
**Pull request by dere-odoo**
Prior to this commit:
When the "Automatic Invoice" setting was enabled and the total amount of the cart was zero due to a discount, no invoice was automatically created in the backend. This required manual intervention to create an invoice from the sales order.

After this commit:
When the "Automatic Invoice" setting is activated and the total amount of the cart is zero thanks to a discount, the invoice is automatically created in the backend without the need for manual operation.

task-3786812

### CODE REVIEW

File: addons/website_sale/controllers/main.py [hunk 1]
```diff
@@ -1768,6 +1768,7 @@ def shop_payment_validate(self, sale_order_id=None, **post):
             order = request.env['sale.order'].sudo().browse(sale_order_id)
             assert order.id == request.session.get('sale_last_order_id')
 
+
```
Comment (kcv-odoo) [inline addons/website_sale/controllers/main.py:1771]:
```suggestion
```
Unnecessary change

File: addons/website_sale/controllers/main.py [hunk 2]
```diff
@@ -1781,6 +1782,14 @@ def shop_payment_validate(self, sale_order_id=None, **post):
 
         if order and not order.amount_total and not tx_sudo:
             order.with_context(send_email=True).with_user(SUPERUSER_ID).action_confirm()
+            auto_invoice = str2bool(
+                    request.env['ir.config_parameter'].sudo().get_param('sale.automatic_invoice')
+                    )
+            if auto_invoice:
```
Comment (kcv-odoo) [inline addons/website_sale/controllers/main.py:1788]: I am not sure but this feels wrong :confused: 
I think what you need to do is change something in actual `auto_invoice` flow and take this into account but as you try to do it was creating payment entry and here we don't want to create payment entries :neutral_face:
Still need dive in deeper :crab:

**Review by kcv-odoo — COMMENTED** [review]
Thanks for your work :100: 

About your solution we still need to explore more about `auto invoice` flow :face_exhaling: : before finalizing solution

File: addons/sale/models/sale_order.py [hunk 3]
```diff
@@ -1295,6 +1295,20 @@ def _get_invoiceable_lines(self, final=False):
 
         return self.env['sale.order.line'].browse(invoiceable_line_ids + down_payment_line_ids)
 
+    def process_zero_value_orders(self):
```
Comment (kcv-odoo) [inline addons/sale/models/sale_order.py:1298]: This should in `website_sale` module also it should be private and it could in `main.py` with order as an arg

File: addons/sale/models/sale_order.py [hunk 4]
```diff
@@ -1295,6 +1295,20 @@ def _get_invoiceable_lines(self, final=False):
...
+        :rtype: bool
+        """
+        self.with_context(send_email=True).action_confirm()
+        auto_invoice = self.env['ir.config_parameter'].get_param('sale.automatic_invoice')
+        if auto_invoice:
+            if not self.amount_to_invoice and not self.amount_paid and self.reward_amount:
+                invoice = self._create_invoices(final=True)
+                invoice.action_post()
+                return True
+        return False
```
Comment (kcv-odoo) [inline addons/sale/models/sale_order.py]:
```suggestion
        if not auto_invoice:
            return
        if not self.amount_to_invoice and not self.amount_paid and self.reward_amount:
            invoice = self._create_invoices(final=True)
            invoice.action_post()
            return True
```

File: addons/sale/models/sale_order.py [hunk 5]
```diff
@@ -1295,6 +1295,24 @@ def _get_invoiceable_lines(self, final=False):
...
+        amount total and creates invoices if necessary.
+
+        :return: True if invoice is generated automatically, False otherwise
+        :rtype: bool
+        """
+        self.action_confirm()
+        auto_invoice = self.env['ir.config_parameter'].get_param('sale.automatic_invoice')
+        if not auto_invoice:
+            return False
+
```
Comment (kcv-odoo) [inline addons/sale/models/sale_order.py:1310]:
```suggestion
```

File: addons/sale/models/sale_order.py [hunk 6]
```diff
@@ -1295,6 +1295,24 @@ def _get_invoiceable_lines(self, final=False):
...
+        :return: True if invoice is generated automatically, False otherwise
+        :rtype: bool
+        """
+        self.action_confirm()
+        auto_invoice = self.env['ir.config_parameter'].get_param('sale.automatic_invoice')
+        if not auto_invoice:
+            return False
+        
+        if not self.amount_to_invoice and not self.amount_paid and self.reward_amount:
```
Comment (kcv-odoo) [inline addons/sale/models/sale_order.py:1311]: comment would great explaining what and why this condition
Same could improve bit better in doc string

Comment (kcv-odoo) [inline addons/sale/models/sale_order.py:1311]:
```suggestion
        if self.reward_amount:
```
Isn't this condition is enough :thinking:

**Review by kcv-odoo — COMMENTED** [review]
Thank you for the changes!
Added some comments

And about adding new method I am not sure moving all things there is a good idea
sorry about that :slightly_frowning_face:

**Review by kcv-odoo — APPROVED** [review]
Look good to me!

File: addons/website_sale/controllers/main.py [hunk 7]
```diff
@@ -1770,6 +1770,13 @@ def shop_payment_validate(self, sale_order_id=None, **post):
 
         if order and not order.amount_total and not tx_sudo:
             order.with_context(send_email=True).with_user(SUPERUSER_ID).action_confirm()
+            auto_invoice = order.env['ir.config_parameter'].get_param('sale.automatic_invoice')
+            # create an invoice for order with zero total amount and automatic invoice enabled
+            if auto_invoice and order.reward_amount:
```
Comment (Feyensv) [inline addons/website_sale/controllers/main.py:1775]: `reward_amount` is defined in `sale_loyalty`, you cannot use that field in `website_sale`.

**Review by Feyensv — CHANGES_REQUESTED** [review]
Could you please add a test to cover this new code ? :pray:

File: addons/website_sale/controllers/main.py [hunk 8]
```diff
@@ -1770,6 +1770,11 @@ def shop_payment_validate(self, sale_order_id=None, **post):
 
         if order and not order.amount_total and not tx_sudo:
             order.with_context(send_email=True).with_user(SUPERUSER_ID).action_confirm()
+            is_order_invoiced = order._invoice_zero_amount_orders()
```
Comment (Feyensv) [inline addons/website_sale/controllers/main.py:1773]: this method doesn't exist in `website_sale` :eyes:

File: addons/website_sale_loyalty/tests/test_website_sale_auto_invoice.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
+from odoo.tests import tagged
+from odoo.fields import Command
+from odoo.addons.website_sale.controllers.main import WebsiteSale
+from odoo.addons.website.tools import MockRequest
+from odoo.addons.website_sale.tests.common import WebsiteSaleCommon
```
Comment (Feyensv) [inline addons/website_sale_loyalty/tests/test_website_sale_auto_invoice.py:5]: Please order your imports alphabetically
```suggestion
# Part of Odoo. See LICENSE file for full copyright and licensing details.

from odoo.fields import Command
from odoo.tests import tagged

from odoo.addons.website.tools import MockRequest
from odoo.addons.website_sale.controllers.main import WebsiteSale
from odoo.addons.website_sale.tests.common import WebsiteSaleCommon
```

File: addons/website_sale_loyalty/tests/test_website_sale_auto_invoice.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+                    'code': "100code",
+                    'minimum_amount': 0,
+                })
+            ],
+            }
+        )
+        reward = self.env['loyalty.reward'].sudo().create({
+            'program_id': program.id,
+            'discount': 100,
+        })
```
Comment (Feyensv) [inline addons/website_sale_loyalty/tests/test_website_sale_auto_invoice.py:35]:
```suggestion
        program = self.env['loyalty.program'].sudo().create({
            'name': '100discount',
            'program_type': 'promo_code',
            'rule_ids': [
                Command.create({
                    'code': "100code",
                    'minimum_amount': 0,
                }),
            ],
            'reward_ids': [
                Command.create({...}
            ],
            }
        )
        reward = program.reward_ids
```
When possible, create the records directly with the complete data, it's more efficient.
I believe that in this case, the program would create a reward automatically, and end up with two rewards, which seems unnecessary for the purpose of your test.

File: addons/website_sale/controllers/main.py [hunk 11]
```diff
@@ -1769,7 +1769,11 @@ def shop_payment_validate(self, sale_order_id=None, **post):
             return request.redirect('/shop')
 
         if order and not order.amount_total and not tx_sudo:
-            order.with_context(send_email=True).with_user(SUPERUSER_ID).action_confirm()
+            is_order_invoiced = order.with_context(send_email=True).with_user(SUPERUSER_ID)._validate_zero_amount_cart()
+            if is_order_invoiced:
+                # clean context and session, then redirect to the portal page
+                request.website.sale_reset()
```
[the remark this answers is not in the document]
Comment (dere-odoo) [reply addons/website_sale/controllers/main.py:1775]: Yes but before i did this task the code did not have `request.website.sale_reset()` and since this task is to address a specific flow (when amount total is zero due to a discount/gift card/...) .I added a condition for `is_order_invoiced`. So as not to break original code flow

[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/website_sale/controllers/main.py:1775]: I believe it's safe and cleaner to do that change.
That way, you wouldn't have to return a boolean value in your new method.

File: addons/website_sale_loyalty/models/sale_order.py [hunk 12]
```diff
@@ -240,3 +240,17 @@ def _cart_find_product_line(self, product_id, line_id=None, **kwargs):
...
+    def _validate_zero_amount_cart(self):
+        """
+        This method overrides the base implementation from the `website_sale` module.
+        Validates the cart when the total amount is zero and creates an invoice if applicable.
+
+        :return: True if an invoice is created and posted, otherwise False.
+        """
+        super()._validate_zero_amount_cart()
+        auto_invoice = self.env['ir.config_parameter'].get_param('sale.automatic_invoice')
+        if auto_invoice and self.reward_amount:
```
Comment (Feyensv) [inline addons/website_sale_loyalty/models/sale_order.py:253]: My bad, I should have noticed earlier, but auto_invoice could be a string value representing False, and it would be True if you check it like that.
You should ideally use `str2bool` util to make sure falsy values won't be considered truthy.

File: addons/website_sale/models/sale_order.py [hunk 13]
```diff
@@ -511,6 +511,16 @@ def _update_cart_line_values(self, order_line, update_values):
         self.ensure_one()
         order_line.write(update_values)
 
+    def _validate_zero_amount_cart(self):
+        """
+        This method is used to validate the cart when the total amount of cart is zero.
+        This method is intended to be overridden by other modules to extend its functionality
+        such as adding automatic invoicing for zero amount orders
+
+        :return: False, when no invoice is generated for this order and True otherwise
```
Comment (Feyensv) [inline addons/website_sale/models/sale_order.py:520]: not returning a bool anymore (and honestly not sure there is a need for a docstring, the code is self-explaining IMHO)

**Comment by Feyensv** [comment]
@robodoo r+ squash
